### PR TITLE
Pod health

### PR DESCRIPTION
**Pull request by alexec**
Reading material:

https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-phase

Looks like this:

Screen Shot 2019-04-03 at 15 44 37

### CODE REVIEW

File: controller/cache/info.go [hunk 1]
```diff
@@ -190,6 +190,29 @@ func populatePodInfo(un *unstructured.Unstructured, node *node) {
 	if reason != "" {
 		node.info = append(node.info, v1alpha1.InfoItem{Name: "Status Reason", Value: reason})
 	}
+	node.health = createHealth(pod.Status, reason)
 	node.info = append(node.info, v1alpha1.InfoItem{Name: "Containers", Value: fmt.Sprintf("%d/%d", readyContainers, totalContainers)})
 	node.networkingInfo = &v1alpha1.ResourceNetworkingInfo{Labels: un.GetLabels()}
 }
+
+func createHealth(status v1.PodStatus, reason string) *v1alpha1.HealthStatus {
```
Comment (jessesuen) [inline controller/cache/info.go]: Can we instead use `health.go::getPodHealth()` exporting it if necessary? If the logic there is not sufficient (e.g. it does not take into consideration `CrashLoopBackOff`), then we should improve that method.

**Comment by jessesuen** [comment]
So I think existing health check for pod should be improved to meet the new requirements we want for pod health.

Regardless if phase is Running/Pending, we still need to check if any  containerStatus' have a bad `state.waiting.reason` set. If yes, it may be Degraded instead of Running/Pending. Examples:

```yaml
  phase: Running
  containerStatuses:
    - containerID: docker://274713675a79687298eef26db4349f1dec87e0222d4ada276d4ef5a44e978e11
      image: XXXX
      imageID: XXXXX
      lastState:
        terminated:
          containerID: >-
            docker://274713675a79687298eef26db4349f1dec87e0222d4ada276d4ef5a44e978e11
          exitCode: 137
          finishedAt: '2019-04-04T07:15:22Z'
          reason: Error
          startedAt: '2019-04-04T07:13:02Z'
      name: icn-routing
      ready: false
      restartCount: 4172
      state:
        waiting:
          message: >-
            Back-off 5m0s restarting failed container=icn-routing
            pod=icn-routing-65d8bd8c4b-wbxvr_identity-relationship-icn-routing-usw2-prd-prd(e0dd84ba-4b82-11e9-9c31-0a42b2f8969e)
          reason: CrashLoopBackOff
```


```yaml
status:
  phase: Pending
  containerStatuses:
    - image: XXXXX
      imageID: ''
      lastState: {}
      name: dgviewer-api
      ready: false
      restartCount: 0
      state:
        waiting:
          message: >-
            Back-off pulling image
            "XXXXX"
          reason: ImagePullBackOff
```

**Comment by jessesuen** [comment]
How can we bring better attention about pods which are in a Running phase but not ready (i.e. what we currently show as `0/1` in very small text).

```yaml
status:
  phase: Running
  containerStatuses:
    - containerID: >-
        docker://5cf4b29d5c82237afb4e0d4ae585ee5e1d32ce457f9136609c3206f92c85530e
      image: 'gcr.io/heptio-images/ks-guestbook-demo:0.2'
      imageID: >-
        docker-pullable://gcr.io/heptio-images/ks-guestbook-demo@sha256:e6fbd640820c32c847a53b93eb7b404f25ffb56df1cc16d0711eec3a405d6509
      lastState: {}
      name: guestbook-ui
      ready: false
      restartCount: 0
      state:
        running:
          startedAt: '2019-04-04T08:18:32Z'
```

File: controller/cache/info.go (hunk 1 above)
Comment (alexec) [reply controller/cache/info.go]: I will investigate

Comment (alexec) [reply controller/cache/info.go]: ok, so both `getPodHealth` and `populatePodInfo` do the same thing and need to be merged

File: util/health/health.go [hunk 2]
```diff
@@ -411,12 +412,32 @@ func getJobHealth(obj *unstructured.Unstructured) (*appv1.HealthStatus, error) {
...
+	var messages []string
+
+	for _, containerStatus := range pod.Status.ContainerStatuses {
+		waiting := containerStatus.State.Waiting
+		// Article listing common container errors: https://medium.com/kokster/debugging-crashloopbackoffs-with-init-containers-26f79e9fb5bf
+		if waiting != nil && (strings.HasPrefix(waiting.Reason, "Err") || strings.HasSuffix(waiting.Reason, "Error") || strings.HasSuffix(waiting.Reason, "BackOff")) {
+			status = appv1.HealthStatusDegraded
+			messages = append(messages, waiting.Message)
+		}
+	}
```
Comment (jessesuen) [inline util/health/health.go]: This logic actually cannot be applied when the pod.Spec.RestartPolicy is: coreV1.RestartPolicyOnFailure, coreV1.RestartPolicyNever, otherwise it breaks the resource hook logic.

The issue is, if we mark a pod with ImagePullBackoff as Degraded, and the pod is used as a resource hook, then we will prematurely fail the PreSync/PostSync hook. Meanwhile, when that error condition is resolved (e.g. the image is available), the resource hook pod will unexpectedly be executed even though the sync has completed.

In other words, whenever pods are run in the context of a job, the Pending and Running phase should always be considered Progressing.

File: util/health/health_test.go [hunk 3]
```diff
@@ -68,6 +68,7 @@ func TestPod(t *testing.T) {
 	assertAppHealth(t, "./testdata/pod-pending.yaml", appv1.HealthStatusProgressing)
 	assertAppHealth(t, "./testdata/pod-running-not-ready.yaml", appv1.HealthStatusProgressing)
 	assertAppHealth(t, "./testdata/pod-crashloop.yaml", appv1.HealthStatusDegraded)
+	assertAppHealth(t, "./testdata/pod-imagepullbackoff.yaml", appv1.HealthStatusDegraded)
 	assertAppHealth(t, "./testdata/pod-error.yaml", appv1.HealthStatusDegraded)
 	assertAppHealth(t, "./testdata/pod-running-restart-always.yaml", appv1.HealthStatusHealthy)
 	assertAppHealth(t, "./testdata/pod-running-restart-never.yaml", appv1.HealthStatusProgressing)
```
Comment (jessesuen) [inline util/health/health_test.go]: Let's add a test where we have a pending pod, with restart policy of Never and OnFailure. The expectation is that the health status is progressing and not degraded.

**Review by jessesuen — DISMISSED** [review]

File: util/health/testdata/pod-pending.yaml [hunk 4]
```diff
@@ -64,10 +64,6 @@ status:
     name: main
     ready: false
     restartCount: 0
-    state:
-      waiting:
-        message: Back-off pulling image "doesnt-exist"
-        reason: ImagePullBackOff
```
Comment (jessesuen) [inline util/health/testdata/pod-pending.yaml]: Rather than remove this, can we replace it with a valid Pending condition where it's waiting but there's no error (e.g. currently waiting to be schedulable)

File: util/health/testdata/pod-running-restart-onfailure.yaml [hunk 5]
```diff
@@ -75,10 +75,6 @@ status:
     name: main
     ready: false
     restartCount: 4
-    state:
-      waiting:
-        message: Back-off 1m20s restarting failed container=main pod=my-pod_argocd(3cf9325e-f617-11e8-a057-fe5f49266390)
-        reason: CrashLoopBackOff
```
Comment (jessesuen) [inline util/health/testdata/pod-running-restart-onfailure.yaml]: We need to keep this state here. The expectation is a pod in this state is Progressing and not Degraded.

File: util/health/testdata/pod-pending.yaml (hunk 4 above)
Comment (jessesuen) [reply util/health/testdata/pod-pending.yaml]: You can use:
```yaml
        waiting:
          reason: PodInitializing
```

**Comment by jessesuen** [comment]
Here is a pod yaml (taken from a real cluster) which is waiting to initialize, but has no problems otherwise. This can be used for pod-pending.yaml  unit test. The pod should be considered Progressing.

```yaml
apiVersion: v1
kind: Pod
metadata:
  creationTimestamp: '2019-03-19T03:29:06Z'
  generateName: my-service-64675b958c-
  name: my-service-64675b958c-bl2g6
  namespace: my-namespace
  ownerReferences:
    - apiVersion: apps/v1
      blockOwnerDeletion: true
      controller: true
      kind: ReplicaSet
      name: my-service-64675b958c
      uid: ad788acb-467c-11e9-a825-0ae1de55dd8e
  resourceVersion: '167391497'
  selfLink: /api/v1/namespaces/my-namespace/pods/my-service-64675b958c-bl2g6
  uid: 26cf704c-49f7-11e9-a825-0ae1de55dd8e
spec:
  containers:
      image: fakeimage
      imagePullPolicy: IfNotPresent
      name: my-service
      ports:
        - containerPort: 8443
          name: service
          protocol: TCP
        - containerPort: 8490
          name: metrics
          protocol: TCP
      resources:
        limits:
          cpu: '1'
          memory: 4096M
        requests:
          cpu: '1'
          memory: 4096M
      terminationMessagePath: /dev/termination-log
      terminationMessagePolicy: File
  dnsPolicy: ClusterFirst
  initContainers:
    - image: fakeimage
      imagePullPolicy: IfNotPresent
      name: segment-app-init
      terminationMessagePath: /dev/termination-log
      terminationMessagePolicy: File
  nodeName: 1.2.3.4
  priority: 0
  restartPolicy: Always
  schedulerName: default-scheduler
  securityContext: {}
  serviceAccount: default
  serviceAccountName: default
  terminationGracePeriodSeconds: 30
  tolerations:
    - effect: NoSchedule
      key: node.kubernetes.io/memory-pressure
      operator: Exists
    - effect: NoExecute
      key: node.kubernetes.io/not-ready
      operator: Exists
      tolerationSeconds: 300
    - effect: NoExecute
      key: node.kubernetes.io/unreachable
      operator: Exists
      tolerationSeconds: 300
    - key: ig.iks.intuit.com/dedicated
      value: nodes
    - key: ig/nodes
status:
  conditions:
    - lastProbeTime: null
      lastTransitionTime: '2019-03-19T03:29:06Z'
      message: 'containers with incomplete status: [segment-app-init]'
      reason: ContainersNotInitialized
      status: 'False'
      type: Initialized
    - lastProbeTime: null
      lastTransitionTime: '2019-03-19T03:29:06Z'
      message: 'containers with unready status: [my-service]'
      reason: ContainersNotReady
      status: 'False'
      type: Ready
    - lastProbeTime: null
      lastTransitionTime: '2019-03-19T03:29:06Z'
      message: 'containers with unready status: [my-service]'
      reason: ContainersNotReady
      status: 'False'
      type: ContainersReady
    - lastProbeTime: null
      lastTransitionTime: '2019-03-19T03:29:06Z'
      status: 'True'
      type: PodScheduled
  containerStatuses:
    - image: fakeimage
      imageID: ''
      lastState: {}
      name: my-service
      ready: false
      restartCount: 0
      state:
        waiting:
          reason: PodInitializing
  hostIP: 10.99.104.68
  initContainerStatuses:
    - image: fakeimage
      imageID: ''
      lastState: {}
      name: segment-app-init
      ready: false
      restartCount: 0
      state:
        waiting:
          reason: PodInitializing
  phase: Pending
  qosClass: Guaranteed
  startTime: '2019-03-19T03:29:06Z'
```

File: util/health/testdata/pod-pending.yaml (hunk 4 above)
Comment (alexec) [reply util/health/testdata/pod-pending.yaml]: I think this is the test case you mention above. I'm going to revert this change and update the fn to match.

**Review by jessesuen — APPROVED** [review]
awesome!
